### ISSUE TITLE
Hello!

### ISSUE THREAD

**Issue by teivah** [SkySoft-ATM/gorillaz#77]
This is an open-source library. So, why can't I take a look 😄?

* There are a couple of lock copies. For example: https://github.com/SkySoft-ATM/gorillaz/blob/427cb8c36de9704842179d644d3ddec88b946c89/sd.go#L110
That's an issue. If a struct has a lock, it should be passed as a pointer. Otherwise, locking will be done on a different object, meaning that's useless as fuck (https://eli.thegreenplace.net/2018/beware-of-copying-mutexes-in-go/).
* There might be some context leaks where the cancel function is not used/passed everywhere. For example: https://github.com/SkySoft-ATM/gorillaz/blob/a504ea87b277d97092097db7f8f99322b27e4bcd/getAndWatch_stream_consumer.go#L141 
If the stream is closed or if there's an error, the context is never ever stopped (which is a leak).
* You guys should use https://godoc.org/golang.org/x/tools/cmd/goimports
* Logging an error is handling an error. It means you should either log an error or returning it (using wrapping as you do sometimes). This rule of thumb is not always followed. As an example: 
https://github.com/SkySoft-ATM/gorillaz/blob/26fc987de171ee60b12222dbad80854aa223cd92/getAndWatch_stream_provider.go#L58
It means you'll have more than one error log leading to the same problem and that's not cool.
* There is a shit load of unhandled error. For example:
https://github.com/SkySoft-ATM/gorillaz/blob/26fc987de171ee60b12222dbad80854aa223cd92/stream_provider.go#L170
* Some parts should be cleaned as they are unused. For example:
https://github.com/SkySoft-ATM/gorillaz/blob/d35f82b8776485e359b56257e087810714c9cbae/mux/state_test.go#L33
Or this `Println` call:
https://github.com/SkySoft-ATM/gorillaz/blob/d35f82b8776485e359b56257e087810714c9cbae/mux/state_test.go#L40 
* If err != nil, else if err != nil
https://github.com/SkySoft-ATM/gorillaz/blob/427cb8c36de9704842179d644d3ddec88b946c89/trace.go#L36
* It's more interesting to pass a string instead of an int in this case:
https://github.com/SkySoft-ATM/gorillaz/blob/41f2a75bd239586490493b517c853c72d2796e1d/pprof.go#L15
The rationale? If you guys want to expose the port on a specific interface, you just pass "myhost:myport" without having to change anything in the doc
* This should not be part of the testing package:
https://github.com/SkySoft-ATM/gorillaz/blob/427cb8c36de9704842179d644d3ddec88b946c89/sd.go#L155
Also, you should take a look at https://godoc.org/github.com/stretchr/testify/mock (as it's already part of the testing package you're using)
* There's a better pattern to handle options instead of passing `interface{}`: https://github.com/ReactiveX/RxGo/blob/v2/options.go
(I did not invent anything, this was already implemented in the gRPC client library)

**Comment by pdaures** [comment]
Hi!
Thanks for your input, we appreciate it!
We took into account several points from your list. First of all, we included `go vet` in our build chain to detect leaking context, and detect issues with locks passed by value instead of reference.

For your point about 
```
 listener, err := net.Listen("tcp", fmt.Sprintf(":%d", serverPort)) 
```
I actually like to have integer as integer, so we can catch issues early when the application start.

This week I'll take care of non handled errors, gosec seems to detect several of them.

I'll add comments in this issue as we move forward.

Thanks for your input!
